### PR TITLE
FEATURE: extract content subgraph inmemory caching

### LINKED ISSUE #5243: Extract Content Subgraph in-memory caching
Currently `ContentGraph::getSubgraph()` returns an instance of `ContentSubgraphWithRuntimeCaches` to increase performance of multiple reads in the same process.
Caching database access on such a low level has drawbacks though, because it adds a lot of complexity related to state management (if runtime caches are not reset properly, this can lead to subtle bugs that are really hard to debug).

Together with @skurfuerst we came up with the following plan:

- [x] Move everything underneath the `...\ContentGraph\ContentGraphWithRuntimeCaches` namespace from the CR core to the `Neos.ContentRepositoryRegistry`
- [x] Rework `InMemoryCache` to be a singleton (e.g. `ContentSubgraphCachePool`) that provides central access to the in-memory cache
- [x] Add `ContentRepositoryRegistry::cachedSubgraphForNode()` that returns the `ContentSubgraphWithRuntimeCaches` instance (with the original subgraph and an the `ContentSubgraphCachePool` injected)
- [x] Use that new method everywhere in the code base where it makes sense (e.g. Eel operations, ...)
- [x] Introduce a CatchUpHook that resets the cached subgraph instances (and the cache pool singleton)
- [x] Rework [NodeController::fillCacheWithContentNodes()](https://github.com/neos/neos-development-collection/blob/afd285f825c681759b4eae51a1f7fe7dc721c9c1/Neos.Neos/Classes/Controller/Frontend/NodeController.php#L305-L373) to use the new singleton and extract the code from the controller such that the implementation can be replaced (or removed if that feature is not needed)

### PR DESCRIPTION
**Pull request by skurfuerst**
Resolves: #5243

### CODE REVIEW

File: Neos.Neos/Classes/Controller/Frontend/NodeController.php [hunk 1]
```diff
@@ -326,22 +322,24 @@ private function fillCacheWithContentNodes(
             self::fillCacheInternal(
                 $childSubtree,
                 $currentDocumentNode,
-                $inMemoryCache
+                $this->subgraphCachePool,
+                $subgraph
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Controller/Frontend/NodeController.php:328]: This `$subgraph` is the *uncached*  version. IMO we should use the cached version here

Comment (skurfuerst) [reply Neos.Neos/Classes/Controller/Frontend/NodeController.php:328]: done

**Review by bwaidelich — APPROVED** [review]
I love it, so many removals and simplifications!
+1 by reading

File: Neos.Neos/Classes/Controller/Frontend/NodeController.php [hunk 2]
```diff
@@ -201,10 +204,11 @@ public function showAction(string $node): void
         }
 
         $contentRepository = $this->contentRepositoryRegistry->get($nodeAddress->contentRepositoryId);
-        $subgraph = $contentRepository->getContentGraph($nodeAddress->workspaceName)->getSubgraph(
+        $uncachedSubgraph = $contentRepository->getContentGraph($nodeAddress->workspaceName)->getSubgraph(
             $nodeAddress->dimensionSpacePoint,
             VisibilityConstraints::frontend()
         );
+        $subgraph = new ContentSubgraphWithRuntimeCaches($uncachedSubgraph, $this->subgraphCachePool);
```
Comment (mhsdesign) [inline Neos.Neos/Classes/Controller/Frontend/NodeController.php:211]: we probably need an api way of doing this as thats sometimes the case on the boundary where we dont have a $node yet and cannot use the cr registry

**Pull request merged**